### ISSUE TITLE
How to close the (detection) accuracy gap compared to the nn.MultiheadAttention in PyTorch

### ISSUE THREAD

**Issue by PkuRainBow** [pytorch/fairseq#3072]
## ❓ Performance drops compared to the nn.MultiheadAttention in PyTorch

 We tried to apply the multi-head attention implementation in this code base
(https://github.com/pytorch/fairseq/blob/a48f235636557b8d3bc4922a6fa90f3a0fa57955/fairseq/modules/multihead_attention.py#L20) to replace the official **nn.MultiheadAttention** while finding the performance drops on the object detection task based on DETR project
(https://github.com/facebookresearch/detr/blob/a54b77800eb8e64e3ad0d8237789fcbf2f8350c5/models/transformer.py#L132)

#### Code

We paste the usage of the official nn.MultiheadAttention as the following:

`self.self_attn = nn.MultiheadAttention(d_model, nhead, dropout=dropout)`

We also modify the fairseq implementation of the MultiheadAttention via:

- NOT use the q_noise & qn_block_size by removing the quant_noise API at (https://github.com/pytorch/fairseq/blob/a48f235636557b8d3bc4922a6fa90f3a0fa57955/fairseq/modules/multihead_attention.py#L65-L77)

- COMMENT the codes at

https://github.com/pytorch/fairseq/blob/a48f235636557b8d3bc4922a6fa90f3a0fa57955/fairseq/modules/multihead_attention.py#L155-L187

#### What have you tried?

We have tried to train our object detection models based the DETR project by simply replacing the official nn.MultiheadAttention with the fairseq implementations while finding the final AP on COCO object detection task drops around 0.5 points from 34.6% to 34.1%.

#### What's your environment?

 - fairseq Version (e.g., 1.0 or master): v0.10.1
 - PyTorch Version (e.g., 1.0): 1.7
 - OS (e.g., Linux): Linux
 - How you installed fairseq (`pip`, source): We do not install fairseq.
 - Python version: 3.7

**Comment by myleott** [comment]
To confirm, when you say performance gap are you referring to training speed or downstream task accuracy?

Assuming you mean downstream task accuracy, my guess is that it's due to the slightly different weight init that fairseq uses: https://github.com/pytorch/fairseq/blob/a48f235636557b8d3bc4922a6fa90f3a0fa57955/fairseq/modules/multihead_attention.py#L100-L104.

The history behind this is we previously used a single input projection layer to go from Q -> Q,K,V, and thus the `fan_out` for `xavier_uniform_` was 3x larger than if you had 3 independent projections. We later reverted to three independent projections, but kept the old weight init for consistency with previous results (and because we observed slightly better performance on some tasks, although I suspect this is task-dependent).

**Comment by PkuRainBow** [comment]
@myleott Thanks for your reply!

**- "To confirm, when you say performance gap are you referring to training speed or downstream task accuracy?"**
We mean the downstream task accuracy on object detection, which is based on Transformer architecture.

**-  "slightly different weight init that fairseq uses:.."**
In fact, I have noticed that this difference and tried to study the performance with "gain=1 / math.sqrt(2)" and "gain=1" while still finding the performance based on fairseq implementation is worse.

**- Other differences**
I also find some other differences in the **_reset_parameters**:

 ```
   def _reset_parameters(self):
        if self._qkv_same_embed_dim:
            xavier_uniform_(self.in_proj_weight)
        else:
            xavier_uniform_(self.q_proj_weight)
            xavier_uniform_(self.k_proj_weight)
            xavier_uniform_(self.v_proj_weight)

        if self.in_proj_bias is not None:
            '''here is different from the fairseq implementations'''
            constant_(self.in_proj_bias, 0.)
            constant_(self.out_proj.bias, 0.)
        if self.bias_k is not None:
            xavier_normal_(self.bias_k)
        if self.bias_v is not None:
            xavier_normal_(self.bias_v)
```
(https://github.com/pytorch/pytorch/blob/b89827b73f7881a8108518452768654894862e5d/torch/nn/modules/activation.py#L912)
